### PR TITLE
[test] Add p4de instance test on SM

### PR DESCRIPTION
**Pull request by saimidu**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Add p4de instance test

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by saimidu** [comment]
Tests run as expected in commit [18c269a](https://github.com/aws/deep-learning-containers/pull/2471/commits/18c269aeb0289655c681a746c4f8afa1a6e10b78), with ml.p4d instances on all tests.

File: test/sagemaker_tests/pytorch/training/__init__.py [hunk 1]
```diff
@@ -11,3 +11,21 @@
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
 from __future__ import absolute_import
+
+import os
+
+
+def get_efa_test_instance_type(default):
```
[the remark this answers is not in the document]
Comment (kace) [reply test/sagemaker_tests/pytorch/training/__init__.py:18]: I guess I'm just weirded out by this function in general. We take a str or list[str] as input and output a list[str], but in some cases it's a list of 1 string that is immediately unpacked into a string.

File: test/sagemaker_tests/pytorch/training/conftest.py [hunk 2]
```diff
@@ -141,10 +144,11 @@ def pytest_configure(config):
...
 def pytest_runtest_setup(item):
-    if item.config.getoption("--efa"):
-        efa_tests = [mark for mark in item.iter_markers(name="efa")]
-        if not efa_tests:
-            pytest.skip("Skipping non-efa tests")
+    efa_tests = [mark for mark in item.iter_markers(name="efa")]
+    if item.config.getoption("--efa") and not efa_tests:
+        pytest.skip("Skipping non-efa tests due to --efa flag")
+    elif not item.config.getoption("--efa") and efa_tests:
+        pytest.skip("Skipping efa tests because --efa flag is missing")
```
Comment (kace) [inline test/sagemaker_tests/pytorch/training/conftest.py:151]: Is it worth logging here in an `else` statement?

File: test/sagemaker_tests/pytorch/training/conftest.py [hunk 3]
```diff
@@ -278,10 +285,11 @@ def fixture_build_base_image(request, framework_version, py_version, processor,
...
     return Session(boto_session=boto3.Session(region_name=region))
 
+
 @pytest.fixture(name='efa_instance_type')
-def fixture_efa_instance_type():
-    default_instance_type = "ml.p3dn.24xlarge"
-    return default_instance_type
+def fixture_efa_instance_type(request):
+    configured_efa_instance_type = get_efa_test_instance_type(default="ml.p4d.24xlarge")
+    return request.param if hasattr(request, "param") else configured_efa_instance_type[0]
```
Comment (kace) [inline test/sagemaker_tests/pytorch/training/conftest.py:292]: Most "Pythonic"/EAFP code here would be:

```
try:
    return request.param
except AttributeError:
    return get_efa_test_instance_type(default="ml.p4d.24xlarge")[0]
```

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py [hunk 4]
```diff
@@ -80,8 +87,19 @@ def invoke_pytorch_estimator(
...
                     continue
                 else:
                     raise e
+
+    instance_types = []
+    if "instance_type" in estimator_parameter:
+        instance_types = [estimator_parameter["instance_type"]]
+    elif "instance_groups" in estimator_parameter:
+        instance_types = [instance_group.instance_type for instance_group in estimator_parameter["instance_groups"]]
+    if "CapacityError" in str(e) and any(instance_type in LOW_AVAILABILITY_INSTANCE_TYPES for instance_type in instance_types):
```
Comment (kace) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py:102]: `str(e)` should be `str(error)` here.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py [hunk 5]
```diff
@@ -11,10 +11,16 @@
...
+
+import time
+
+import pytest
 import sagemaker
+
 from sagemaker.pytorch import PyTorch
-import time
+
+from ..... import get_ecr_image, get_ecr_image_region, get_sagemaker_session, LOW_AVAILABILITY_INSTANCE_TYPES
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py:22]: Yeah, I'm not really a fan of relative imports. But we've already started doing it, and this is a problem that needs to be solved in multiple places. I can try to see how I can change this to better local imports.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py [hunk 6]
```diff
@@ -65,21 +67,21 @@ def test_smdataparallel_throughput(framework_version, ecr_image, sagemaker_regio
             "iterations": 100,
             "warmup": 10,
             "bucket_size": 25,
-            "info": "PT-{}-N{}".format(instance_types, 2)
+            "info": "PT-{}-N{}".format(efa_instance_type, 2),
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:70]: Makes sense. I'll change it here. Thanks!

File: test/test_utils/__init__.py [hunk 7]
```diff
@@ -25,6 +25,7 @@
 # from security import EnhancedJSONEncoder
 
 from src import config
+from src.config import are_efa_tests_enabled
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/test_utils/__init__.py:28]: Actually I can remove this completely. I think this was leftover after a code change that I reverted.

File: test/sagemaker_tests/pytorch/training/conftest.py (hunk 2 above)
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/conftest.py:151]: This is two separate boolean variables that need to be true in "xnor".

File: test/sagemaker_tests/__init__.py [hunk 8]
```diff
@@ -16,6 +16,9 @@
 import re
 from test.test_utils.ecr import reupload_image_to_test_ecr
 
+LOW_AVAILABILITY_INSTANCE_TYPES = ["ml.p4de.24xlarge"]
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/sagemaker_tests/__init__.py:19]: I added the explanatory comment where the list is used. Will add it here as well.

**Review by kace — APPROVED** [review]

**Pull request merged**